File: crawler/src/transform.ts
import {
	assertFileExists,
	readJsonFile,
	resolveInputPath,
	resolveOutputPath,
	writeJsonFile,
} from "./utils/files.ts";
import type {
	GroupBase,
	Message,
	TransformInputPayload,
	TransformOutputPayload,
} from "./utils/types.ts";

const MEGA_LINK_REGEX = /https?:\/\/(?:www\.)?mega\.(?:nz|co\.nz)\/[^\s)>]+/gi;
const URL_REGEX = /https?:\/\/[^\s)>]+/gi;
const CIRCLE_REGEX = /^\s*\*\*(.+?)\*\*(?:\s+.*)?$/;
const STATUS_LINE_REGEX = /^\s*Stat(?:us|s)(?:\s*\([^)]*\))?\s*:\s*(.+)$/i;
const MISSING_LINE_REGEX = /^\s*Missing(?:\s*\([^)]*\))?\s*:/i;
const STATUS_UPDATE_REGEX = /-\s*(?:Last\s+)?Updat\w*\s*:?\s*(.+)$/i;
const TRAILING_DATE_REGEX = /-\s*(\d{1,2}\/\d{1,2}\/\d{2,4})\s*$/;

type MutableGroup = {
	links: Set<string>;
	missingLink: string | null;
	status: string | null;
	statusMeta: string | null;
	lastUpdated: string | null;
};

function normalizeCircleName(rawName: string): string {
	return rawName.trim();
}

function normalizeStatus(rawStatus: string): string | null {
	const normalized = rawStatus
		.replace(/^[:\-\s]+/, "")
		.replace(/\s+/g, " ")
		.trim();

	if (normalized.length === 0) {
		return null;
	}

	const compact = normalized.toLowerCase().replace(/[^a-z]/g, "");

	if (compact.startsWith("incomplete") || compact.startsWith("incompleted")) {
		return "incomplete";
	}

	if (compact.startsWith("missing")) {
		return "missing";
	}

	if (
		compact.startsWith("completed") ||
		compact.startsWith("complete") ||
		compact.endsWith("completed") ||
		compact.startsWith("compelted")
	) {
		return "completed";
	}

	return normalized;
}

function parseStatusLine(line: string): {
	status: string | null;
	statusMeta: string | null;
	lastUpdated: string | null;
} | null {
	const match = line.match(STATUS_LINE_REGEX);
	if (!match) {
		return null;
	}

	const value = match[1].trim();
	const updatedMatch = value.match(STATUS_UPDATE_REGEX) ?? value.match(TRAILING_DATE_REGEX);
	const lastUpdated = updatedMatch?.[1].trim() ?? null;

	const withoutUpdate = updatedMatch
		? value.slice(0, Math.max(0, updatedMatch.index ?? value.length)).trim()
		: value;

	const metaMatch = withoutUpdate.match(/\[([^\]]+)\]/);
	const statusMeta = metaMatch?.[1].trim() ?? null;

	const withoutMeta = withoutUpdate.replace(/\[[^\]]+\]/g, " ");
	const status = normalizeStatus(withoutMeta.replace(/_/g, " "));

	return { status, statusMeta, lastUpdated };
}

function getOrCreateGroup(groups: Map<string, MutableGroup>, circle: string): MutableGroup {
	const existing = groups.get(circle);
	if (existing) {
		return existing;
	}

	const created: MutableGroup = {
		links: new Set<string>(),
		missingLink: null,
		status: null,
		statusMeta: null,
		lastUpdated: null,
	};

	groups.set(circle, created);
	return created;
}

function collectGroups(messages: Message[]): GroupBase[] {
	const groups = new Map<string, MutableGroup>();

	for (const message of messages) {
		const content = typeof message.content === "string" ? message.content : "";
		if (!content) {
			continue;
		}

		let currentCircle: string | null = null;
		const lines = content.split(/\r?\n/);

		for (const line of lines) {
			const circleMatch = line.match(CIRCLE_REGEX);
			if (circleMatch) {
				const normalized = normalizeCircleName(circleMatch[1]);
				currentCircle = normalized.length > 0 && normalized !== " " ? normalized : null;

				if (currentCircle) {
					getOrCreateGroup(groups, currentCircle);
				}
			}

			if (!currentCircle) {
				continue;
			}

			const group = getOrCreateGroup(groups, currentCircle);

			const parsedStatus = parseStatusLine(line);
			if (parsedStatus) {
				if (parsedStatus.status) {
					group.status = parsedStatus.status;
				}

				if (parsedStatus.statusMeta) {
					group.statusMeta = parsedStatus.statusMeta;
				}

				if (parsedStatus.lastUpdated) {
					group.lastUpdated = parsedStatus.lastUpdated;
				}
			}

			for (const match of line.matchAll(MEGA_LINK_REGEX)) {
				group.links.add(match[0]);
			}

			if (MISSING_LINE_REGEX.test(line) && group.missingLink === null) {
				group.missingLink = line.match(URL_REGEX)?.at(0) ?? null;
			}
		}
	}

	return [...groups.entries()]
		.map(([circle, group]) => ({
			circle,
			links: [...group.links],
			missingLink: group.missingLink,
			status: group.status,
			statusMeta: group.statusMeta,
			lastUpdated: group.lastUpdated,
		}))
		.filter((group) => group.links.length > 0);
}

export async function transform(inputArg?: string, outputArg?: string): Promise<void> {
	const inputPath = resolveInputPath(inputArg, "dist/input.json");
	const outputPath = resolveOutputPath(outputArg, "dist/transformed.json");

	await assertFileExists(inputPath);

	const inputJson = await readJsonFile<TransformInputPayload>(inputPath);

	if (!Array.isArray(inputJson.messages)) {
		throw new Error("Invalid input JSON: expected top-level 'messages' array.");
	}

	const groups = collectGroups(inputJson.messages as Message[]);
	const outputJson: TransformOutputPayload = { groups };

	await writeJsonFile(outputPath, outputJson);

	const totalLinks = groups.reduce((sum, group) => sum + group.links.length, 0);
	console.log(`Wrote ${groups.length} circles and ${totalLinks} MEGA links to ${outputPath}`);
}

File: cli/internal/pipeline/transform/transform.go
package transform

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	"github.com/Curstantine/dcidxr/cli/internal/files"
	"github.com/Curstantine/dcidxr/cli/internal/model"
)

var (
	megaLinkPattern     = regexp.MustCompile(`https?://(?:www\.)?mega\.(?:nz|co\.nz)/[^\s)>]+`)
	urlPattern          = regexp.MustCompile(`https?://[^\s)>]+`)
	circlePattern       = regexp.MustCompile(`^\s*\*\*(.+?)\*\*(?:\s+.*)?$`)
	statusLinePattern   = regexp.MustCompile(`^\s*Stat(?:us|s)(?:\s*\([^)]*\))?\s*:\s*(.+)$`)
	missingLinePattern  = regexp.MustCompile(`^\s*Missing(?:\s*\([^)]*\))?\s*:`)
	statusUpdatePattern = regexp.MustCompile(`-\s*(?:Last\s+)?Updat\w*\s*:?\s*(.+)$`)
	trailingDatePattern = regexp.MustCompile(`-\s*(\d{1,2}/\d{1,2}/\d{2,4})\s*$`)
)

type transformInputPayload struct {
	Messages []model.Message `json:"messages"`
}

type mutableGroup struct {
	Links       []string
	seenLinks   map[string]struct{}
	MissingLink *string
	Status      *string
	StatusMeta  *string
}

func Run(_ context.Context, inputArg, outputArg string) error {
	inputPath := files.ResolvePath(inputArg, "dist/input.json")
	outputPath := files.ResolvePath(outputArg, "dist/transformed.json")

	inputJSON, err := files.ReadJSONFile[transformInputPayload](inputPath)
	if err != nil {
		return err
	}

	if inputJSON.Messages == nil {
		return fmt.Errorf("invalid input JSON: expected top-level 'messages' array")
	}

	groups := collectGroups(inputJSON.Messages)
	output := model.TransformOutputPayload{Groups: groups}

	if err := files.WriteJSONFile(outputPath, output); err != nil {
		return err
	}

	totalLinks := 0
	for _, group := range groups {
		totalLinks += len(group.Links)
	}

	fmt.Printf("Wrote %d circles and %d MEGA links to %s\n", len(groups), totalLinks, outputPath)
	return nil
}

func collectGroups(messages []model.Message) []model.GroupBase {
	groups := map[string]*mutableGroup{}
	order := make([]string, 0)

	for _, message := range messages {
		content, ok := message.Content.(string)
		if !ok || content == "" {
			continue
		}

		var currentCircle string
		for _, line := range strings.Split(content, "\n") {
			circleMatch := circlePattern.FindStringSubmatch(line)
			if len(circleMatch) > 1 {
				normalized := normalizeCircleName(circleMatch[1])
				if normalized != "" && normalized != " " {
					currentCircle = normalized
					if _, ok := groups[currentCircle]; !ok {
						groups[currentCircle] = &mutableGroup{
							Links:     []string{},
							seenLinks: map[string]struct{}{},
						}
						order = append(order, currentCircle)
					}
				}
			}

			if currentCircle == "" {
				continue
			}

			group := groups[currentCircle]
			status, statusMeta, _ := parseStatusLine(line)
			if status != nil {
				group.Status = status
			}
			if statusMeta != nil {
				group.StatusMeta = statusMeta
			}

			for _, match := range megaLinkPattern.FindAllString(line, -1) {
				if _, exists := group.seenLinks[match]; exists {
					continue
				}
				group.seenLinks[match] = struct{}{}
				group.Links = append(group.Links, match)
			}

			if missingLinePattern.MatchString(line) && group.MissingLink == nil {
				urls := urlPattern.FindAllString(line, -1)
				if len(urls) > 0 {
					link := urls[0]
					group.MissingLink = &link
				}
			}
		}
	}

	out := make([]model.GroupBase, 0, len(order))
	for _, circle := range order {
		group := groups[circle]
		links := append([]string(nil), group.Links...)

		if len(links) == 0 {
			continue
		}

		out = append(out, model.GroupBase{
			Circle:      circle,
			Links:       links,
			MissingLink: group.MissingLink,
			Status:      group.Status,
			StatusMeta:  group.StatusMeta,
		})
	}

	return out
}

func normalizeCircleName(raw string) string {
	return strings.TrimSpace(raw)
}

func normalizeStatus(raw string) *string {
	normalized := strings.TrimSpace(raw)
	normalized = strings.TrimLeft(normalized, " :-")
	normalized = strings.Join(strings.Fields(normalized), " ")
	if normalized == "" {
		return nil
	}

	compact := strings.ToLower(regexp.MustCompile(`[^a-z]`).ReplaceAllString(normalized, ""))

	value := normalized
	switch {
	case strings.HasPrefix(compact, "incomplete"), strings.HasPrefix(compact, "incompleted"):
		value = "incomplete"
	case strings.HasPrefix(compact, "missing"):
		value = "missing"
	case strings.HasPrefix(compact, "completed"), strings.HasPrefix(compact, "complete"), strings.HasSuffix(compact, "completed"), strings.HasPrefix(compact, "compelted"):
		value = "completed"
	}

	return &value
}

func parseStatusLine(line string) (status *string, statusMeta *string, lastUpdated *string) {
	match := statusLinePattern.FindStringSubmatch(line)
	if len(match) < 2 {
		return nil, nil, nil
	}

	value := strings.TrimSpace(match[1])
	updatedMatch := statusUpdatePattern.FindStringSubmatchIndex(value)
	if updatedMatch == nil {
		updatedMatch = trailingDatePattern.FindStringSubmatchIndex(value)
	}

	withoutUpdate := value
	if updatedMatch != nil {
		if len(updatedMatch) >= 4 {
			lu := strings.TrimSpace(value[updatedMatch[2]:updatedMatch[3]])
			if lu != "" {
				lastUpdated = &lu
			}
		}

		start := updatedMatch[0]
		if start >= 0 && start <= len(value) {
			withoutUpdate = strings.TrimSpace(value[:start])
		}
	}

	metaPattern := regexp.MustCompile(`\[([^\]]+)\]`)
	metaMatch := metaPattern.FindStringSubmatch(withoutUpdate)
	if len(metaMatch) > 1 {
		meta := strings.TrimSpace(metaMatch[1])
		if meta != "" {
			statusMeta = &meta
		}
	}

	withoutMeta := metaPattern.ReplaceAllString(withoutUpdate, " ")
	status = normalizeStatus(strings.ReplaceAll(withoutMeta, "_", " "))

	return status, statusMeta, lastUpdated
}
